src/cargo/core/resolver/mod.rs

```diff
@@ -298,109 +244,213 @@ fn activate(mut cx: Box<Context>,
-    // each one in turn.
-    let mut last_err = None;
-    for candidate in my_candidates {
         trace!("{}[{}]>{} trying {}", parent.name(), cur, dep.name(),
                candidate.version());
-        let mut my_cx = cx.clone();
-        my_cx.resolve.graph.link(parent.package_id().clone(),
-                                 candidate.package_id().clone());
+        cx.resolve.graph.link(parent.package_id().clone(),
+                              candidate.package_id().clone());
 
         // If we hit an intransitive dependency then clear out the visitation
         // list as we can't induce a cycle through transitive dependencies.
         if !dep.is_transitive() {
-            my_cx.visited.clear();
-        }
-        let my_cx = try!(activate(my_cx, registry, candidate, &method,
-                                  &mut |cx, registry| {
-            activate_deps(cx, registry, parent, platform, deps.clone(), cur + 1,
-                          finished)
-        }));
-        match my_cx {
-            Ok(cx) => return Ok(Ok(cx)),
-            Err(e) => { last_err = Some(e); }
+            cx.visited.clear();
         }
+        remaining_deps.extend(
+            try!(activate(&mut cx, registry, candidate, &method)));
     }
-    trace!("{}[{}]>{} -- {:?}", parent.name(), cur, dep.name(), last_err);
+    debug!("resolved: {:?}", cx.resolve);
+    Ok(cx.resolve)
+}
 
-    // Oh well, we couldn't activate any of the candidates, so we just can't
-    // activate this dependency at all
-    Ok(activation_error(&cx, registry, last_err, parent, dep, prev_active,
-                        &candidates))
+// Searches up `backtrack_stack` until it finds a dependency with remaining
+// candidates. Resets `cx` and `remaining_deps` to that level and returns the
+// next candidate. If all candidates have been exhausted, returns an activation
+// error.
+fn find_candidate(backtrack_stack: &mut Vec<BacktrackFrame>,
+                  cx: &mut Context, remaining_deps: &mut RemainingDeps,
+                  registry: &mut Registry,
+                  mut last_err: Box<CargoError>) -> CargoResult<Rc<Summary>> {
+    while let Some(mut frame) = backtrack_stack.pop() {
+        let maybe_candidate = match frame.remaining_candidates.next() {
+            None => {
+                trace!("{}[{}]>{} -- {:?}", frame.parent.name(), frame.cur, frame.dep.name(),
+                       Some(&last_err));
+                last_err = activation_error(cx, registry, Some(last_err), &frame.parent, &frame.dep,
+                                            &frame.prev_active, &frame.all_candidates);
+                None
+            }
+            Some((_, candidate)) => {
+                *cx = frame.context_backup.clone();
+                *remaining_deps = frame.deps_backup.clone();
+                Some(candidate.clone())
+            }
+        };
+        if let Some(candidate) = maybe_candidate {
+            backtrack_stack.push(frame);
```

Comment:
`frame.remaining_candidates.next()` borrows `frame` because of the non-Iterator `next()` signature. It might make sense to have this `next()` function auto-clone its result, since I think I'm explicitly cloning all the uses now.